### DISCUSSION TITLE
Using workspaces in docker builds

### DISCUSSION THREAD

**Discussion by nad-au** [oven-sh/bun#12763] [General]
I'm looking to go monorepo and use https://bun.sh/docs/install/workspaces but I'm unsure how to build docker images when the app has one or more workspace dependencies. Currently our dockerfiles are structured as per https://bun.sh/guides/ecosystem/docker

Is anyone doing docker builds from workspaces? Care to share a dockerfile?

**Comment by handlebauer** [comment]
For anyone else looking to solve this still, compare your approach to:

Project structure:
```bash
$ tree -L 3 -I 'node_modules'
.
├── README.md
├── bun.lockb
├── package.json
├── packages
│   ├── pkg-a
│   │   ├── Dockerfile
│   │   ├── README.md
│   │   ├── package.json
│   │   ├── src
│   │   └── tsconfig.json
│   └── pkg-b
│       ├── README.md
│       ├── package.json
│       ├── src
│       └── tsconfig.json
└── tsconfig.json
```

Dockerfile: `./packages/pkg-a/Dockerfile`
```Dockerfile
# Use the official Bun image
FROM oven/bun:1

# Set working directory
WORKDIR /app

# Copy package.json files
COPY package.json bun.lockb ./
COPY packages/pkg-a/package.json ./packages/pkg-a/
COPY packages/pkg-b/package.json ./packages/pkg-b/

# Install dependencies
RUN bun install --frozen-lockfile

# Copy the rest of the application
COPY . .

# Set the entrypoint to run pkg-a
CMD ["bun", "run", "packages/pkg-a/src/index.ts"]
```

Build image:
```bash
docker build -f packages/pkg-a/Dockerfile -t docker-bun-workspace .
```

Workspace config mirrors that found @ https://bun.sh/guides/install/workspaces

**Reply by nad-au** [reply]
Thank you so much @handlebauer I'll be sure to try this

**Reply by handlebauer** [reply]
Some quick pointers:
- You might encounter an issue during the `RUN bun install --frozen-lockfile` build step if you have e.g. 5 packages in your workspace and you don't `COPY` the `package.json`s for all of them (even if you're building an image that requires deps from 4 or less those of packages; this was fine for me but could be a dealbreaker depending on context)
- Another `RUN bun install --frozen-lockfile` related issue may occur unless you downgrade (`FROM oven/bun:1.0.35` worked for me, as per https://github.com/oven-sh/bun/issues/12919#issuecomment-2256448185)

**Comment by janos-r** [comment]
1) Doesn't this install all the dependencies of all the packages, just to run one package?
2) Isn't it the same as fist `COPY . .` and then `bun install`?

**Comment by handlebauer** [comment]
I was looking into this quite a while ago now but to your first point, yes, installing deps for all packages was the workaround (at the time; not sure if applicable still). Not 100% on your second point but I see your point and... probably, yes?

**Reply by janos-r** [reply]
Based on https://dev.to/is_bik/how-create-bun-workspaces-and-build-it-with-docker-51c4
it seams to me that these days one can build just for one app with `bun install --filter 'pkg-a'`
Looking into it now and considering if to start a monorepo with bun or deno.

**Comment by nad-au** [comment]
Just returning to this as we're working on a large monorepo (~10 apps + ~30 libs) and indeed we're effectively shipping the entire monorepo on each app build. While this is fine, the source files are minimal in size, and we're sharing a lot of the libs across our effective distributed monolith, I'm looking for a more optimal solution.

We've separately been trialling [single-file executables](https://bun.sh/docs/bundler/executables) and wondering if it would be OK to ship an executable as a single file to a Docker image? Currently a compiled build is running at ~110Mb. And a single file vs ~30,000 files would speed up builds and docker image size.

Anyone else tried executables in docker images?

**Comment by handlebauer** [comment]
Neat workaround. Sometimes I want to `docker exec -it` into a container and debug the source in that environment which I guess you'd forfeit in this case but outside of that I don't see any obvious downsides.

**Comment by jsemldonado** [comment]
Filed a feature request for `bun prune --docker` — basically turborepo's prune but native to bun: https://github.com/oven-sh/bun/issues/28600

In the meantime, the cleanest Docker pattern I've found for bun workspaces:

```dockerfile
# syntax=docker/dockerfile:1
COPY package.json bun.lock ./
COPY --parents services/*/package.json packages/*/package.json ./
RUN bun install --frozen-lockfile --filter '@myapp/api'

COPY packages/ packages/
COPY services/api/ services/api/
RUN bun run --filter '@myapp/api' build
```

`COPY --parents` (BuildKit) auto-globs all workspace manifests in one line — no manual list to maintain. `--filter` on both install and run scopes to one workspace. `bun build --target=bun` tree-shakes unused packages so you can copy all of `packages/` without worrying about it.

**Comment by AriyanCastle** [comment]
One quick note, add `**/node_modules` to your `.dockerignore` in the root, this is critical, if you don't, monorepo symlinks will be over written and you'll get errors like `ENOENT looking for /app/packages/pkg-a/node_modules` while it should be looking for `/app/node_modules`

**Comment by codemaster012345678910** [comment]
thank you for the info

**Comment by codemaster012345678910** [comment]
this is new to me as i have joined only a couple of months ago
